### ISSUE TITLE
Should fully stick to C++11 memory model

### ISSUE THREAD

**Issue by wks** [microvm/microvm-meta#4]
Currently the memory ordering primitives are copied from LLVM. As stated by LLVM langref, their memory model is not precisely defined.

> These semantics (ordering) are borrowed from Java and C++0x, but are somewhat more colloquial. If these descriptions aren’t precise enough, check those specs (see spec references in the atomics guide).

But the MicroVM should have a precise memory model. The C++11 memory model is the result of a lot of effort and we should stick to it.

Things should be done in the MicroVM:
- Remove the "UNORDERED" memory order. It is designed by LLVM for Java, but the MicroVM will treat data race as undefined behaviour and will leave the security constraints of Java to the client.
  - If JVM can be implemented in C/C++, its memory model should also be implementable on a MicroVM with C++11-like memory model.
- Add "CONSUME" memory order. Define the "carries-a-data-dependency-to" relation in MicroVM. Some hardwares are aware of dependency.
- Define the program order (which is a total order per MicroVM thread because MicroVM does not have unspecified parameter evaluation order), the synchronisation order, the synchronises-with and the happens-before relations.

**Comment by wks** [comment]
Some note about the differences between C and µVM:
- The "sequenced before" ordering in C (the counterpart of "program order" in Java) is a partial order because of the unspecified order of function argument evaluation. In µVM, however, the program order should be a **total order** because the client generate codes that evaluate parameters and µVM simply does exactly what the client requests. (Haskell evaluates arguments lazily. The client has to take care of it.)
- The C programming language defines "memory location" as the region of storage (the "object") of either a scalar value or a sequence of bit fields which do not have 0 size. The **µVM does not have bit fields**, hence the "memory location" is equivalent to the concept of "object" in C, which is defined as the "region of storage". So the µVM specification will use "memory location" all the time and avoid using "object" (which, in the µVM context, means a "heap object").
- The µVM itself **does not provide locks** as the C standard library does. The µVM will permit locks to be designed by library programmer.
- The µVM garbage collector **may move objects**. The "memory location" should be preserved when an object is moved. That is, a specific offset in an object is considered the same location even when the object is moved.
- The µVM has memory allocation as part of its instruction set. The writes of the _initial values_ by NEW/ALLOCA/global should be strong enough for multi-threaded use, but weak enough not to hamper efficient execution.

**Comment by wks** [comment]
# About the "consume" order.

## "consume" and RCU

"Consume" is weaker than "acquire". [McKenney et al. pointed out](http://www2.rdrop.com/users/paulmck/RCU/consume.2014.05.26e.pdf) that It does not require a memory fence on platforms including ARM, Itanium and PowerPC because the hardware tracks the data dependency internally. "Consume" is the memory order needed to implement [RCU](http://lwn.net/Articles/262464/). The cost of the fence introduced by "acquire" may be amplified when RCU is used to traverse a list, in which case every iteration requires an "acquire" or "consume" load.

The µVM has a built-in garbage collector which makes RCU more efficient and easier to implement. RCU-style updating is also the style of functional programming. I expect high-level languages that make extensive use of immutable types (including Scala and some Java programs) can greatly benefit from this.

Since the µVM hardly do any optimisation to the code, it should be designed to mirror the hardware behaviour.

## Criticisms

The main problem with "consume" is that it is not currently implemented well in main-stream compilers. 

[Linus Torvalds criticised](http://www.realworldtech.com/forum/?threadid=143222&curpostid=143417) "consume" for not being implemented properly.

[This GCC bug](https://gcc.gnu.org/bugzilla/show_bug.cgi?id=59448) also shows some problem in current implementations.

[This lwn.net article](http://lwn.net/Articles/588300/) states that the problem of "consume" is "that it requires extensive tracking of dependencies between data accesses", which is difficult to understand or implement. It also mentioned that some compilers simply translate it to "acquire". Some compiler transformations may eliminate data dependencies and such transformations are undesirable.

## Architecture-specific Notes:

### POWER

Excerpt from PowerISA spec v2.06b, section 1.7.1

Data dependency:

> If a Load instruction depends on the value returned by a preceding Load instruction (because the
> value is used to compute the effective address specified by the second Load), the corresponding storage accesses are performed in program order with respect to any processor or mechanism to the extent required by the associated Memory Coher- ence Required attributes. This applies even if the dependency has no effect on program logic (e.g., the value returned by the first Load is ANDed with zero and then added to the effective address spec- ified by the second Load).

Control dependency:

> Because processors may predict branch target addresses and branch condition resolution, control dependencies (e.g., branches) do not order stor- age accesses except as described above. For example, when a subroutine returns to its caller the return address may be predicted, with the result that loads caused by instructions at or after the return address may be performed before the load that obtains the return address is performed.

### ARMv7

Excerpt from ARMv7ar manual section A3.8.2

Dependency:

> Address dependency
> 
> An address dependency exists when the value returned by a read access is used for the computation of the virtual address of a subsequent read or write access. An address dependency exists even if the value read by the first read access does not change the virtual address of the second read or write access. This might be the case if the value returned is masked off before it is used, or if it has no effect on the predicted address value for the second access.
> 
> Control dependency
> 
> A control dependency exists when the data value returned by a read access determines the condition flags, and the values of the flags are used in the condition code checking that determines the address of a subsequent read access. This address determination might be through conditional execution, or through the evaluation of a branch.

Ordering:

> - If there is an address dependency then the two memory accesses are observed in program order by any observer in the common shareability domain of the two accesses.
> - This ordering restriction does not apply if there is only a control dependency between the two read accesses.
> - If there is both an address dependency and a control dependency between two read accesses the ordering requirements of the address dependency apply.

**Comment by wks** [comment]
# About the transformations the µVM can do.

_Abstract: It does not make sense to simply forbid µVM from doing certain transformations. The µVM is always doing transformation, from µVM IR to machine code. The µVM should instead provide enough mechanism so that undesired behaviours (like "speculative writes" mentioned below) can be prevented from being observed in multi-threaded programs with external entities._

According to [this article](http://lwn.net/Articles/586838/), the Linux kernel developers are very unhappy about some compiler optimisations which appears to be harmless to applications.

One of such transformations is **speculative writes**. Because of memory-mapped IO and the multi-threaded nature of the kernel, speculative writes will introduce bugs in the kernel. I believe it will affect multi-threaded programs, too.

An excerpt from [Committee Draft — April 12, 2011 ISO/IEC 9899:201x](http://www.open-std.org/jtc1/sc22/WG14/www/docs/n1570.pdf):

> NOTE 13 Compiler transformations that introduce assignments to a potentially shared memory location that would not be modified by the abstract machine are **generally** precluded by this standard, since such an assignment might overwrite another assignment by a different thread in cases in which an abstract machine execution would not have encountered a data race. This includes implementations of data member assignment that overwrite adjacent members in separate memory locations. We also generally preclude reordering of atomic loads in cases in which the atomics in question may alias, since this may violate the "visible sequence" rules.

**Non-local optimisations** can also be dangerous. In the kernel context, the hardware may update a memory variable which a whole-program analyser may find to be constant. There are also entities that are not the µVM program, including the Client and other libraries or processes, that update the memory in parallel to µVM programs. The µVM should be aware of those external entities.

In conclusion, the µVM memory model should be designed in such a way that when the used properly, the following transformations (by the µVM itself or the underlying CPU) can be made illegal:
- removing or reordering memory access instructions.
- moving memory access instructions across function calls or external actions.
- speculative writes that are visible to other threads or external entities.

**Comment by wks** [comment]
A page for the memory model is made available. https://github.com/microvm/microvm-spec/wiki/memory-model
